api/types/saml.go

```diff
@@ -383,165 +374,109 @@ func (o *SAMLConnectorV2) GetTraitMappings() TraitMappingSet {
 	}
 	if o.Spec.AssertionConsumerService == "" {
-		return nil, trace.BadParameter("missing acs - assertion consumer service parameter, set service URL that will receive POST requests from SAML")
+		return trace.BadParameter("missing acs - assertion consumer service parameter, set service URL that will receive POST requests from SAML")
 	}
 	if o.Spec.ServiceProviderIssuer == "" {
 		o.Spec.ServiceProviderIssuer = o.Spec.AssertionConsumerService
 	}
 	if o.Spec.Audience == "" {
 		o.Spec.Audience = o.Spec.AssertionConsumerService
 	}
-	certStore := dsig.MemoryX509CertificateStore{
-		Roots: []*x509.Certificate{},
-	}
+
 	// if we have a entity descriptor url, fetch it first
 	if o.Spec.EntityDescriptorURL != "" {
 		resp, err := http.Get(o.Spec.EntityDescriptorURL)
 		if err != nil {
-			return nil, trace.Wrap(err)
+			return trace.Wrap(err)
 		}
 		if resp.StatusCode != http.StatusOK {
-			return nil, trace.BadParameter("status code %v when fetching from %q", resp.StatusCode, o.Spec.EntityDescriptorURL)
+			return trace.BadParameter("status code %v when fetching from %q", resp.StatusCode, o.Spec.EntityDescriptorURL)
 		}
 		defer resp.Body.Close()
 		body, err := ioutil.ReadAll(resp.Body)
 		if err != nil {
-			return nil, trace.Wrap(err)
+			return trace.Wrap(err)
 		}
 		o.Spec.EntityDescriptor = string(body)
 		log.Debugf("[SAML] Successfully fetched entity descriptor from %q", o.Spec.EntityDescriptorURL)
 	}
 	if o.Spec.EntityDescriptor != "" {
-		metadata := &types.EntityDescriptor{}
+		// metadata is a slimmed down anonymous struct of
+		// github.com/russellhaering/gosaml2/types.EntityDescriptor
+		// to avoid unnecessary imports.
+		metadata := &struct {
+			EntityID         string `xml:"entityID,attr"`
+			IDPSSODescriptor *struct {
+				SingleSignOnServices []struct {
+					XMLName  xml.Name `xml:"urn:oasis:names:tc:SAML:2.0:metadata SingleSignOnService"`
+					Location string   `xml:"Location,attr"`
+				} `xml:"SingleSignOnService"`
+			} `xml:"IDPSSODescriptor,omitempty"`
+		}{}
+
 		err := xml.Unmarshal([]byte(o.Spec.EntityDescriptor), metadata)
 		if err != nil {
-			return nil, trace.Wrap(err, "failed to parse entity_descriptor")
+			return trace.Wrap(err, "failed to parse entity_descriptor")
 		}
 
-		for _, kd := range metadata.IDPSSODescriptor.KeyDescriptors {
-			for _, samlCert := range kd.KeyInfo.X509Data.X509Certificates {
-				certData, err := base64.StdEncoding.DecodeString(strings.TrimSpace(samlCert.Data))
-				if err != nil {
-					return nil, trace.Wrap(err)
-				}
-				cert, err := x509.ParseCertificate(certData)
-				if err != nil {
-					return nil, trace.Wrap(err, "failed to parse certificate in metadata")
-				}
-				certStore.Roots = append(certStore.Roots, cert)
-			}
-		}
 		o.Spec.Issuer = metadata.EntityID
 		if len(metadata.IDPSSODescriptor.SingleSignOnServices) > 0 {
 			o.Spec.SSO = metadata.IDPSSODescriptor.SingleSignOnServices[0].Location
 		}
 	}
```

Comment:
In `CheckAndSetDefaults` check if `EntityDescriptorURL` or `EntityDescriptor` is set, move the rest of the logic to `services.GetSAMLServiceProvider` which is where it was before right?